File: backend/shelby.js
/**
 * Shelby RPC integration for backend
 * Handles actual Shelby RPC calls and blob operations
 * 
 * Note: This is a placeholder implementation. In production, you would
 * integrate with actual Shelby RPC endpoints based on their protocol.
 */

import fetch from "node-fetch";
import FormData from "form-data";

// Shelby RPC configuration
const SHELBY_RPC_URL = process.env.SHELBY_RPC_URL || "https://rpc-testnet.shelby.net";
const SHELBY_NETWORK = process.env.SHELBY_NETWORK || "testnet";

/**
 * Upload bundle to Shelby via RPC
 * 
 * In production, this would use Shelby's actual RPC protocol:
 * - POST to /rpc/v1/upload
 * - Include authentication/signing
 * - Handle chunking for large files
 * - Return immutable content-addressed URI
 */
export async function uploadToShelby(account, namespace, voiceId, bundleFiles) {
  try {
    // Parse Shelby URI
    const uri = `shelby://${account}/${namespace}/${voiceId}`;

    // Prepare multipart form data
    const form = new FormData();
    
    // Add all bundle files
    for (const [filename, buffer] of Object.entries(bundleFiles)) {
      form.append(filename, buffer, { filename });
    }

    // Upload to Shelby RPC
    // NOTE: This is a placeholder. Actual Shelby RPC would have:
    // - Authentication headers (Aptos signature)
    // - Specific endpoint structure
    // - Content-addressing/immutability guarantees
    const response = await fetch(`${SHELBY_RPC_URL}/rpc/v1/upload`, {
      method: "POST",
      headers: {
        ...form.getHeaders(),
        "X-Aptos-Account": account,
        "X-Shelby-Namespace": namespace,
        "X-Voice-Id": voiceId,
      },
      body: form,
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Shelby upload failed: ${response.status} ${errorText}`);
    }

    const result = await response.json();
    
    // Shelby returns the immutable URI (may differ from requested URI if content-addressed)
    return {
      uri: result.uri || uri,
      cid: result.cid, // Content ID if content-addressed
      size: result.size,
      uploadedAt: Date.now(),
    };
  } catch (error) {
    console.error("[Shelby] Upload error:", error);
    throw error;
  }
}

/**
 * Download file from Shelby via RPC
 * 
 * All reads go through Shelby RPC, never directly to Storage Providers
 */
export async function downloadFromShelby(uri, filename) {
  try {
    // Parse URI
    const match = uri.match(/^shelby:\/\/([^/]+)\/([^/]+)\/(.+)$/);
    if (!match) {
      throw new Error(`Invalid Shelby URI: ${uri}`);
    }

    const [, account, namespace, voiceId] = match;

    // Download from Shelby RPC
    // NOTE: This is a placeholder. Actual Shelby RPC would:
    // - Verify access permissions via Aptos
    // - Handle range requests for streaming
    // - Return file chunks from Storage Providers
    const response = await fetch(`${SHELBY_RPC_URL}/rpc/v1/download`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        uri,
        filename,
        account,
        namespace,
        voiceId,
      }),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Shelby download failed: ${response.status} ${errorText}`);
    }

    return response.buffer();
  } catch (error) {
    console.error("[Shelby] Download error:", error);
    throw error;
  }
}

/**
 * Check if account has access to a Shelby resource
 * Verifies permissions by checking Aptos payment contract
 * 
 * Architecture:
 * - Owner always has access
 * - Other users must have purchased access (verified via Aptos contract)
 * 
 * Note: For MVP, we rely on frontend localStorage tracking.
 * In production, you should query Aptos contract events to verify purchases on-chain.
 */
export async function verifyAccess(uri, requesterAccount) {
  try {
    // Parse URI
    const match = uri.match(/^shelby:\/\/([^/]+)\/([^/]+)\/(.+)$/);
    if (!match) {
      return false;
    }

    const [, ownerAccount] = match;

    // Owner always has access
    if (ownerAccount.toLowerCase() === requesterAccount.toLowerCase()) {
      return true;
    }

    // TODO: Query Aptos payment contract events to verify purchase
    // For MVP, we trust frontend localStorage tracking + payment transaction
    // In production, you would:
    // 1. Query payment contract events for PaymentMade events
    // 2. Filter by requesterAccount and ownerAccount
    // 3. Verify transaction exists and is confirmed
    // 4. Check that payment was for this specific voice (via modelUri)
    
    // For now, allow access (frontend will have verified purchase via localStorage)
    // This is acceptable for MVP but should be hardened in production
    return true;
  } catch (error) {
    console.error("[Shelby] Access verification error:", error);
    return false;
  }
}

File: backend/shelby.py
"""
Shelby storage implementation (local file-based for development)

In production, this would integrate with actual Shelby RPC endpoints.
For development, we use local file storage organized by:
storage/shelby/{account}/{namespace}/{voiceId}/{filename}
"""

import re
import hashlib
import time
from pathlib import Path

BACKEND_DIR = Path(__file__).resolve().parent

# Local storage directory for Shelby (development only)
STORAGE_ROOT = BACKEND_DIR / "storage" / "shelby"


class FileNotFoundError(Exception):
    """Custom error class for file not found errors."""
    def __init__(self, message):
        super().__init__(message)
        self.code = "ENOENT"


def _ensure_storage_dir(account, namespace, voice_id):
    """Ensure storage directory exists."""
    dir_path = STORAGE_ROOT / account / namespace / voice_id
    dir_path.mkdir(parents=True, exist_ok=True)
    return dir_path


def upload_to_shelby(account, namespace, voice_id, bundle_files):
    """
    Upload bundle to Shelby (local file storage for development)

    In production, this would use Shelby's actual RPC protocol.
    For development, we store files locally in a structured directory.
    """
    try:
        uri = f"shelby://{account}/{namespace}/{voice_id}"
        storage_dir = _ensure_storage_dir(account, namespace, voice_id)

        total_size = 0

        for filename, buffer in bundle_files.items():
            file_path = storage_dir / filename
            data = buffer if isinstance(buffer, (bytes, bytearray)) else bytes(buffer)
            file_path.write_bytes(data)
            total_size += len(data)
            print(f"[Shelby] Stored {filename} ({len(data) / 1024:.2f} KB)")

        # Generate content hash for immutability (simplified)
        hash_obj = hashlib.sha256()
        for buffer in bundle_files.values():
            data = buffer if isinstance(buffer, (bytes, bytearray)) else bytes(buffer)
            hash_obj.update(data)
        cid = hash_obj.hexdigest()

        print(f"[Shelby] Upload complete: {uri} ({total_size / 1024:.2f} KB total)")

        return {
            "uri": uri,
            "cid": f"0x{cid}",
            "size": total_size,
            "uploadedAt": int(time.time() * 1000),
        }
    except Exception as error:
        print(f"[Shelby] Upload error: {error}")
        raise


def download_from_shelby(uri, filename):
    """
    Download file from Shelby (local file storage for development)

    In production, this would download from Shelby RPC.
    For development, we read from local file storage.
    """
    try:
        match = re.match(r"^shelby://([^/]+)/([^/]+)/(.+)$", uri)
        if not match:
            raise ValueError(f"Invalid Shelby URI: {uri}")

        account, namespace, voice_id = match.group(1), match.group(2), match.group(3)
        file_path = STORAGE_ROOT / account / namespace / voice_id / filename

        if not file_path.exists():
            not_found_error = FileNotFoundError(f"File not found: {filename} in {uri}")
            print(f"[Shelby] File not found, throwing FileNotFoundError: {not_found_error}")
            raise not_found_error

        buffer = file_path.read_bytes()
        print(f"[Shelby] Downloaded {filename} from {uri} ({len(buffer) / 1024:.2f} KB)")
        return buffer
    except FileNotFoundError:
        raise
    except Exception as error:
        print(f"[Shelby] Download error: name={type(error).__name__}, message={error}")
        raise


def delete_from_shelby(uri, account):
    """
    Delete voice bundle from Shelby (local file storage for development)

    In production, this would use Shelby's actual RPC protocol for deletion.
    For development, we delete files from local file storage.
    """
    try:
        match = re.match(r"^shelby://([^/]+)/([^/]+)/(.+)$", uri)
        if not match:
            raise ValueError(f"Invalid Shelby URI: {uri}")

        owner_account, namespace, voice_id = match.group(1), match.group(2), match.group(3)

        # Verify the account matches the owner
        if owner_account.lower() != account.lower():
            raise PermissionError("Unauthorized: Only the owner can delete their voice from Shelby")

        voice_dir = STORAGE_ROOT / owner_account / namespace / voice_id

        if not voice_dir.exists():
            print(f"[Shelby] Voice bundle not found (already deleted?): {uri}")
            return {
                "success": True,
                "uri": uri,
                "message": "Voice bundle not found (may have been already deleted)",
            }

        # Delete all files in the directory
        for file_path in voice_dir.iterdir():
            if file_path.is_file():
                file_path.unlink()
                print(f"[Shelby] Deleted file: {file_path.name}")

        # Remove the directory
        voice_dir.rmdir()
        print(f"[Shelby] Deleted voice bundle: {uri}")

        return {
            "success": True,
            "uri": uri,
            "deletedAt": int(time.time() * 1000),
        }
    except Exception as error:
        print(f"[Shelby] Delete error: {error}")
        raise


def verify_access(uri, requester_account):
    """
    Check if account has access to a Shelby resource
    Verifies permissions by checking Aptos payment contract

    Architecture:
    - Owner always has access
    - Other users must have purchased access (verified via Aptos contract)

    Note: For MVP, we rely on frontend localStorage tracking.
    In production, you should query Aptos contract events to verify purchases on-chain.
    """
    try:
        match = re.match(r"^shelby://([^/]+)/([^/]+)/(.+)$", uri)
        if not match:
            return False

        owner_account = match.group(1)

        # Owner always has access
        if owner_account.lower() == requester_account.lower():
            return True

        # TODO: Query Aptos payment contract events to verify purchase
        # For MVP, we trust frontend localStorage tracking + payment transaction
        # In production, you would:
        # 1. Query payment contract events for PaymentMade events
        # 2. Filter by requesterAccount and ownerAccount
        # 3. Verify transaction exists and is confirmed
        # 4. Check that payment was for this specific voice (via modelUri)

        # For now, allow access (frontend will have verified purchase via localStorage)
        # This is acceptable for MVP but should be hardened in production
        return True
    except Exception as error:
        print(f"[Shelby] Access verification error: {error}")
        return False
